update-trust-anchors.sh: do not append new trust to existing tls-ca-bundle-client.pem

jouvin opened the pull request:
- Previous behavior considered buggy: a lot of entries duplicated, bundle unnecessarily big
- Variable `CA_BUNDLE_APPEND_NEW_TRUST` can be defined to 1 to retain the old behaviour: remove once assessing it is not useful at all.
  - CERN confirmed in Mattermost it was not relying on this buggy feature

This PR also adds a variable (`REMOVE_OBSOLETE_FILES_FROM_TARGET`) to remove from targets (`CA_BUNDLE_TARGET` and `TRUST_ANCHORS_TARGET`)  files not present in sources.

jouvin committed:
update-trust-anchors.sh: do not append new trust to existing tls-ca-bundle-client.pem

- Previous behavior considered buggy: a lot of entries duplicated,
  bundle unnecessarily big
- Variable CA_BUNDLE_APPEND_NEW_TRUST can be defined to 1 to retain the old
  behaviour

jouvin committed:
update-trust-anchors.sh: add an option to remove obsolete files from targets

- Set variable REMOVE_OBSOLETE_FILES_FROM_TARGET to 1 (0 by default)
- Applies to both CA_BUNDLE_TARGET and TRUST_ANCHORS_TARGET

jouvin committed:
update-trust-anchors.sh: fix CA capitalization in messages

jouvin committed:
update-trust-anchors.sh: remove duplicated code

- Use the first script parameter as the TRUST_ANCHORS_TARGET default value

jouvin commented:
@giacomini this PR should be ready for review. I let you confirm it is ok and merge it.

giacomini commented on update-trust-anchors.sh:
```diff
@@ -1,6 +1,16 @@
+
+# Variable to remove from CA_BUNDLE_TARGET files no longer existing in /etc/pki
+# Disabled by default (backward compatibility)
+REMOVE_OBSOLETE_FILES_FROM_TARGET=${CA_BUNDLE_REMOVE_OBSOLETE_FILES:-0}
```
```suggestion
REMOVE_OBSOLETE_FILES_FROM_TARGET=${REMOVE_OBSOLETE_FILES_FROM_TARGET:-0}
```

jouvin replied:
Good catch :-) I'll fix it.

jouvin committed:
Remove variable CA_BUNDLE_APPEND_NEW_TRUST

- Makes no sense to preserve the old buggy behaviour which is just a source of problems...

jouvin commented:
@giacomini  mispelling of variable name (after renaming it) fixed. I also decided to add a commit to remove the variable allowing to preserve the old buggy behaviour as the more I think about it, the more I consider it is a non-sense and doesn't bring any added value, just problems (when the file becomes too big).

jouvin commented:
@giacomini it is another forgotten PR... Any good reason not to merge it. It is **really critical* for us...

giacomini commented on README.md:
```diff
@@ -38,7 +38,8 @@ volume is mounted (where the container /etc/grid-security/certificates is rsync'
 the bundle `tls-ca-bundle-all.pem` that contains the usual system bundle plus the egi trust-anchors
 - `CA_BUNDLE_SECRET_TARGET` (default unset): an optional kubernetes secret including tls-ca-bundle-all.pem
-
+- `REMOVE_OBSOLETE_FILES_FROM_TARGET` (default: 0): if 1, remove files that are no longer present in the source
+from `CA_BUNDLE_TARGET` and `TRUST_ANCHORS_TARGET`.
```
```suggestion
- `REMOVE_OBSOLETE_FILES_FROM_TARGET` (default unset): if set, remove from the targets (`CA_BUNDLE_TARGET`, `CA_BUNDLE_SECRET_TARGET` and `TRUST_ANCHORS_TARGET`) CA files that are no longer present in the source
```
For uniformity with other control variables, I would use set/unset. We can remove the explicit list of targets, so that we don't need to change this documentation if we add another target. BTW, is this option relevant for the Java keystore?

giacomini replied:
In fact, the CA_BUNDLE_SECRET_TARGET is created from scratch and replaces the previous one. I would like confirmation, though. @balciiberk what do you think?

jouvin replied:
Agreed

balciiberk replied:
Yes, it does that because if we just `kubectl apply` which edits the existing secret, than the old config is put in the annotation `kubectl.kubernetes.io/last-applied-configuration` and as our secret is huge, it is bigger than the allowed limit for this annotation. That's why I configured it to create a new one every time.

But in any case,it already behaves as you expect, and the old ones would automatically be removed.

giacomini commented on update-trust-anchors.sh:
```diff
@@ -1,6 +1,11 @@
 
+# Variable to remove from CA_BUNDLE_TARGET files no longer existing in /etc/pki
+# Disabled by default (backward compatibility)
+REMOVE_OBSOLETE_FILES_FROM_TARGET=${REMOVE_OBSOLETE_FILES_FROM_TARGET:-0}
```
```suggestion
# Variable to remove from the targets the CA files no longer existing in /etc/pki
# Disabled by default (backward compatibility)
REMOVE_OBSOLETE_FILES_FROM_TARGET=${REMOVE_OBSOLETE_FILES_FROM_TARGET:=}
```

jouvin replied:
Agreed

giacomini commented on update-trust-anchors.sh:
```diff
@@ -18,28 +23,38 @@ done
-## Update ca trust does not include trust anchors that can sign client-auth certs,
-## which looks like a bug
+## Updated CA trust does not include trust anchors that can sign client-auth certs,
+## which looks like a bug: readd it after extracting the new trust.
```
```suggestion
## update-ca-trust does not include trust anchors that can sign client-auth certs,
## which looks like a bug: re-add them after extracting the new trust.
```

jouvin replied:
Agreed

giacomini commented on update-trust-anchors.sh:
```diff
@@ -18,28 +23,38 @@ done
 CA_BUNDLE_TARGET=${CA_BUNDLE_TARGET:=}
 
+if [ ${REMOVE_OBSOLETE_FILES_FROM_TARGET} -eq 1 ]
+then
```
```suggestion
if [ ${REMOVE_OBSOLETE_FILES_FROM_TARGET} -eq 1 ]; then
```
For uniformity with the rest of the code

jouvin replied:
Fine

giacomini commented on update-trust-anchors.sh:
```diff
@@ -18,28 +23,38 @@ done
+  delete_msg='and removing obsolete file from it'
+else
+  delete_options=''
+  delete_msg=''
```
```suggestion
```
Are these necessary? if a variable is used without being defined it is empty, isn't i?

jouvin replied:
A matter of preference, I admit. Personally I tend to prefer initialized variables...

giacomini commented on update-trust-anchors.sh:
```suggestion
```
This is not needed any more, right?

jouvin replied:
Not sure what it was intended to do... I kept it for backward compatibility as it doesn't hurt but I don't know if some people depends on this... We don't!

giacomini replied:
In fact it is needed. See comment above about setting `TRUST_ANCHORS_TARGET` to `$1`

jouvin commented:
@giacomini as said in a private conversatin, I let you  take care of this final cleanup, ok?

giacomini commented on update-trust-anchors.sh:
```diff
@@ -18,28 +23,34 @@ done
-TRUST_ANCHORS_TARGET=${TRUST_ANCHORS_TARGET:=}
+# For backward compatibility, allow to define TRUST_ANCHORS_TARGET as a script parameter rather
+# than defining the variable
+TRUST_ANCHORS_TARGET=${TRUST_ANCHORS_TARGET:=$1}
```
```suggestion
TRUST_ANCHORS_TARGET=${TRUST_ANCHORS_TARGET:=}
```
It's not the same thing. At the moment one can set the variable `TRUST_ANCHORS_TARGET` *and* pass a different command-line parameter. With the proposed modification the trust anchors would be copied twice in the target directory.

giacomini commented on update-trust-anchors.sh:
```suggestion
  echo "Copying trust anchors to $1 ${delete_msg}"
```

giacomini commented on update-trust-anchors.sh:
```diff
@@ -84,6 +93,6 @@ if [ -n "${JAVA_BUNDLE_TARGET}" ]; then
 if [ $# -gt 0 ]; then
-  echo "Certificate copy requested to $1"
+  echo "Copying trust anchors to $1 ${delete_msg}"
   rsync -avu -O --no-owner --no-group --no-perms /etc/grid-security/certificates/ $1
```
```suggestion
  rsync -avu ${delete_options} -O --no-owner --no-group --no-perms /etc/grid-security/certificates/ $1
```
